compiler/src/iree/compiler/Codegen/LLVMGPU/AMDGPUChainedMatmulPass.cpp

```diff
@@ -82,71 +85,104 @@ struct AMDGPUPrepareForChainedMatmulPass
+    auto findPermutedMap = [&perm](AffineMap M, AffineMap S) -> AffineMap {
+      LLVM_DEBUG(llvm::dbgs() << "M:\n"; M.print(llvm::dbgs());
+                 llvm::dbgs() << "\n"; llvm::dbgs() << "S:\n";
+                 S.print(llvm::dbgs()); llvm::dbgs() << "\n";);
+      // X(S) = M
+      // X:
+      SmallVector<int64_t> X;
+      for (AffineExpr s : S.getResults()) {
+        for (auto [i, m] : llvm::enumerate(M.getResults())) {
+          if (s == m) {
+            X.push_back(i);
+            break;
+          }
+        }
+      }
 
-  /// For a matmul_transpose_b, this transformation boils down to an operand
-  /// swap and result transpose:
-  ///
-  /// def matmul_transpose_b(A, B):
-  ///   B.T = transpose(B)
-  ///   C = A @ B.T
-  ///   return C
-  ///
-  /// def matmul_transpose_b_swapped(A, B):
-  ///   A.T = transpose(A)
-  ///   C.T = B @ A.T
-  ///   C   = transpose(C.T)
-  ///   return C
-  ///
-  /// matmul_transpose_b(B, A) = matmul_transpose_b_swapped(B, A).T
-  ///
-  /// For the sake of completeness, we also show that this does not hold
-  /// for normal matmul:
-  ///
-  /// def matmul(A, B):
-  ///   C = A @ B
-  ///   return C
-  ///
-  /// def matmul_swapped(A, B):
-  ///  A.T = transpose(A)
-  ///  B.T = transpose(B)
-  ///  C.T = B.T @ A.T
-  ///  C   = transpose(C.T)
-  ///
-  /// TODO: This check applies more generally when one of the operands in the
-  /// function is transposed compared to what "@" expects.
-  bool isOperandSwapInvariant(vector::ContractionOp contractOp) const {
-    AffineExpr m, n, k;
-    bindDims(contractOp.getContext(), m, n, k);
-    using MapList = ArrayRef<ArrayRef<AffineExpr>>;
-    auto infer = [&](MapList m) {
-      return AffineMap::inferFromExprList(m, contractOp.getContext());
+      LLVM_DEBUG(llvm::dbgs() << "X:\n"; interleaveComma(X, llvm::dbgs());
+                 llvm::dbgs() << "\n";);
+
+      // X_T = X(T):
+      applyPermutationToVector(X, perm);
+
+      LLVM_DEBUG(llvm::dbgs() << "X(T):\n"; interleaveComma(X, llvm::dbgs());
+                 llvm::dbgs() << "\n";);
+
+      auto xt = AffineMap::getPermutationMap(X, M.getContext());
+      return xt;
     };
-    SmallVector<AffineMap> newIndexingMaps = infer({{m, k}, {n, k}, {m, n}});
-    return newIndexingMaps == contractOp.getIndexingMapsArray();
+
+    MLIRContext *ctx = rewriter.getContext();
+    AffineMap lhsMap = indexingMaps[0];
+    AffineMap rhsMap = indexingMaps[1];
+    AffineMap accMap = indexingMaps[2];
+
+    LLVM_DEBUG(llvm::dbgs() << "Old Contraction Maps:\n";
+               lhsMap.print(llvm::dbgs()); llvm::errs() << "\n";
+               rhsMap.print(llvm::dbgs()); llvm::errs() << "\n";
+               accMap.print(llvm::dbgs()); llvm::errs() << "\n";);
+
+    AffineExpr m, n, k;
+    bindDims(ctx, m, n, k);
+
+    auto lhsStandardMap = AffineMap::get(3, 0, {m, k}, ctx);
+    auto rhsStandardMap = AffineMap::get(3, 0, {k, n}, ctx);
+    auto accStandardMap = AffineMap::get(3, 0, {m, n}, ctx);
```

Comment:
Does this work for contractions with more than 3 dimensions? Based on my read of the below indexing map code it looks like not. Can you add a guard to this function or earlier in this pass that restricts to variants of a standard MNK matmul?